packages/flutter/lib/src/services/text_editing_value.dart

```diff
@@ -0,0 +1,326 @@
+    int? lastNonWhitespace;
+    for (final String currentString in string.characters) {
+      if (!includeWhitespace &&
+          !isWhitespace(currentString.characters.first.codeUnitAt(0))) {
+        lastNonWhitespace = count;
+      }
+      if (count + currentString.length >= index) {
+        return includeWhitespace ? count : lastNonWhitespace ?? 0;
+      }
+      count += currentString.length;
+    }
+    return 0;
+  }
+
+  /// Creates a copy of this value but with the given fields replaced with the new values.
+  TextEditingValue copyWith({
+    String? text,
+    TextSelection? selection,
+    TextRange? composing,
+  }) {
+    return TextEditingValue(
+      text: text ?? this.text,
+      selection: selection ?? this.selection,
+      composing: composing ?? this.composing,
+    );
+  }
+
+  /// Whether the [composing] range is a valid range within [text].
+  ///
+  /// Returns true if and only if the [composing] range is normalized, its start
+  /// is greater than or equal to 0, and its end is less than or equal to
+  /// [text]'s length.
+  ///
+  /// If this property is false while the [composing] range's `isValid` is true,
+  /// it usually indicates the current [composing] range is invalid because of a
+  /// programming error.
+  bool get isComposingRangeValid => composing.isValid && composing.isNormalized && composing.end <= text.length;
+
+  // Deletes the current non-empty selection.
+  //
+  // If the selection is currently non-empty, this method deletes the selected
+  // text. Otherwise this method does nothing.
+  TextEditingValue _deleteNonEmptySelection() {
+    assert(selection.isValid);
+    assert(!selection.isCollapsed);
+
+    final String textBefore = selection.textBefore(text);
+    final String textAfter = selection.textAfter(text);
+    final TextSelection newSelection = TextSelection.collapsed(
+      offset: selection.start,
+      affinity: selection.affinity,
+    );
+    final TextRange newComposingRange = !composing.isValid || composing.isCollapsed
+      ? TextRange.empty
+      : TextRange(
+        start: composing.start - (composing.start - selection.start).clamp(0, selection.end - selection.start),
+        end: composing.end - (composing.end - selection.start).clamp(0, selection.end - selection.start),
+      );
+
+    return TextEditingValue(
+      text: textBefore + textAfter,
+      selection: newSelection,
+      composing: newComposingRange,
+    );
+  }
+
+  /// Deletes to the given index, inclusively.
+  ///
+  /// Returns a new TextEditingValue representing the state after the deletion.
+  ///
+  /// If the selection is not collapsed, deletes the selection regardless of the
+  /// given index.
+  TextEditingValue deleteTo(int index) {
+    assert(selection != null);
+
+    if (!selection.isValid) {
+      return this;
+    }
+
+    if (!selection.isCollapsed) {
+      return _deleteNonEmptySelection();
+    }
+
+    final String textBefore = selection.textBefore(text);
```

Comment:
There seems to be code duplication, 243-260 looks very similar to 262 - 281. This part of the implementation will probably be cleaner if we first figure out the normalized `TextRange` to delete, and then use "start" and "end" instead of "index" and "selection.extentOffset".